pkg/sdkserver/sdkserver.go

```diff
@@ -486,9 +469,53 @@ func (s *SDKServer) WatchGameServer(_ *sdk.Empty, stream sdk.SDK_WatchGameServer
 }
 
 // Reserve moves this GameServer to the Reserved state for the Duration specified
-// TODO: implement this functionality
-func (s *SDKServer) Reserve(_ context.Context, d *sdk.Duration) (*sdk.Empty, error) {
-	return &sdk.Empty{}, errors.New("not implemented")
+func (s *SDKServer) Reserve(ctx context.Context, d *sdk.Duration) (*sdk.Empty, error) {
+	s.stopReserveTimer()
+
+	e := &sdk.Empty{}
+
+	s.logger.Info("Received Reserve request, adding to queue")
+	s.enqueueState(stablev1alpha1.GameServerStateReserved)
+
+	// 0 is forever.
+	if d.Seconds == 0 {
+		return e, nil
+	}
+
+	duration := time.Duration(d.Seconds) * time.Second
+	s.gsUpdateMutex.Lock()
+	s.gsReserveDuration = &duration
+	defer s.gsUpdateMutex.Unlock()
+
+	return e, nil
+}
+
+// resetReserveAfter will run a function after duration to move the GameServer
+// back to being Ready
+func (s *SDKServer) resetReserveAfter(ctx context.Context, duration time.Duration) {
+	s.gsUpdateMutex.Lock()
+	defer s.gsUpdateMutex.Unlock()
+	if s.reserveTimer != nil {
+		s.reserveTimer.Stop()
+	}
+
+	s.reserveTimer = time.AfterFunc(duration, func() {
+		_, err := s.Ready(ctx, &sdk.Empty{})
```

Comment:
nit: The if should be "cuddled":

```go
if _, err := s.Ready(...); err != nil {
  s.logger.WithError(...)
}
```

to reduce the scope of the `err` variable.